backend/msgpack/queue_mgr.go

```diff
@@ -0,0 +1,155 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/m3db/m3metrics/protocol/msgpack"
+	"github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/pool"
+)
+
+var (
+	errInstanceQueueManagerClosed = errors.New("instance queue manager closed")
+)
+
+// instanceQueueManager manages instance queues.
+type instanceQueueManger interface {
+	// AddInstances adds instances.
+	AddInstances(instances []instance) error
+
+	// RemoveInstances removes instances.
+	RemoveInstances(instances []instance) error
+
+	// WriteTo writes a msgpack encoded buffer to the target instances.
+	WriteTo(instances []instance, buf msgpack.Buffer) error
+
+	// Close closes the connection manager.
+	Close() error
+}
+
+type queueManager struct {
+	sync.RWMutex
+
+	opts        ServerOptions
+	writeOpPool *writeOpPool
+	queues      map[string]*refCountedQueue
+	closed      bool
+}
+
+func newInstanceQueueManager(opts ServerOptions) instanceQueueManger {
+	writeOpPoolOpts := pool.NewObjectPoolOptions().SetSize(opts.WriteOpPoolSize())
+	writeOpPool := newWriteOpPool(writeOpPoolOpts)
+	writeOpPool.Init()
+
+	return &queueManager{
+		opts:        opts,
+		writeOpPool: writeOpPool,
+		queues:      make(map[string]*refCountedQueue),
+	}
+}
+
+func (mgr *queueManager) AddInstances(instances []instance) error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	if mgr.closed {
+		return errInstanceQueueManagerClosed
+	}
+	for _, instance := range instances {
+		id := instance.ID()
+		queue, exists := mgr.queues[id]
+		if !exists {
+			queue = newRefCountedQueue(instance, mgr.opts)
+			mgr.queues[id] = queue
+		}
+		queue.IncRef()
+	}
+	return nil
+}
+
+func (mgr *queueManager) RemoveInstances(instances []instance) error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	if mgr.closed {
+		return errInstanceQueueManagerClosed
+	}
+	for _, instance := range instances {
+		id := instance.ID()
+		queue, exists := mgr.queues[id]
+		if !exists {
+			continue
+		}
+		if queue.DecRef() == 0 {
+			delete(mgr.queues, id)
+		}
+	}
+	return nil
+}
+
+func (mgr *queueManager) WriteTo(instances []instance, buf msgpack.Buffer) error {
+	mgr.RLock()
+	if mgr.closed {
+		mgr.RUnlock()
+		return errInstanceQueueManagerClosed
+	}
+	if len(instances) == 0 {
+		mgr.RUnlock()
+		buf.Close()
+		return nil
+	}
+	multiErr := xerrors.NewMultiError()
+	writeOp := mgr.writeOpPool.Get()
+	writeOp.Reset(len(instances), buf)
+	for _, instance := range instances {
+		queue, exists := mgr.queues[instance.ID()]
+		if !exists {
+			writeOp.DecRef()
+			err := fmt.Errorf("no instance queue exists for %s", instance.String())
+			multiErr = multiErr.Add(err)
+			continue
+		}
+		if err := queue.Enqueue(writeOp); err != nil {
+			writeOp.DecRef()
```

Comment:
I understand why we have ref counts for queues, but why have the ref count for writeOps?